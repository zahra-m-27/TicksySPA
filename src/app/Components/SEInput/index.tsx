import {Tag} from 'antd';
import Assets from '../../Assets';
import styles from './styles.module.scss';
import React, {useEffect, useState} from 'react';
import ClassNames from '../../Utilities/ClassNames';

interface Props {
  hint?: string;
  label?: string;
  regex?: RegExp;
  tags?: string[];
  minLines?: number;
  content?: string;
  className?: string;
  hasError?: boolean;
  icon?: JSX.Element;
  isNumeric?: boolean;
  onEnter?: () => void;
  attachments?: File[];
  inputClassName?: string;
  labelClassName?: string;
  type?: 'password' | 'text';
  onIconPressed?: () => void;
  passwordCanBeVisible?: boolean;
  innerContainerClassName?: string;
  onSelectFile?: (file: File) => void;
  onTagClose?: (index: number) => void;
  onRemoveAttachment?: (index: number) => void;
  onChangeText?: (text: string, HasError: boolean) => void;
}

const SEInput = React.forwardRef<HTMLTextAreaElement | HTMLInputElement, Props>(
  (
    {
      icon,
      tags,
      regex,
      label,
      content,
      onEnter,
      minLines,
      className,
      hint = '',
      onTagClose,
      attachments,
      onSelectFile,
      onChangeText,
      type = 'text',
      onIconPressed,
      labelClassName,
      inputClassName,
      hasError = false,
      isNumeric = false,
      onRemoveAttachment,
      innerContainerClassName,
      passwordCanBeVisible = true,
    }: Props,
    ref
  ) => {
    const [Type, setType] = useState(type);
    const [HasError, setHasError] = useState(false);
    const [IsFocused, setIsFocused] = useState(false);
    const [Content, setContent] = useState(content ?? '');

    useEffect(() => {
      setContent(content ?? '');
    }, [content]);

    useEffect(() => {
      setHasError(hasError);

      if (hasError) {
        setIsFocused(true);
      }
    }, [hasError]);

    const onFocus = () => {
      setIsFocused(true);
    };

    const onBlur = () => {
      if (Content.length <= 0) {
        setIsFocused(false);
      }
    };

    const onChange = (event: React.ChangeEvent<any>) => {
      if (isNumeric && !/^\d*$/.test(event.target.value)) {
        setContent(event.target.value);
        setTimeout(() => {
          setContent('');
          setHasError(true);
        }, 150);
        return;
      }

      if (regex && !regex.test(event.target.value)) {
        setHasError(true);
      } else if (HasError) {
        setHasError(false);
      }

      setContent(event.target.value);
      onChangeText && onChangeText(event.target.value, HasError);
    };

    let inputContainerStyle = ClassNames(styles.input_container, className);
    if (IsFocused) {
      if (HasError) {
        inputContainerStyle = ClassNames(
          styles.input_container,
          styles.focused_input_container,
          styles.failed_regex_focused_input_container,
          className
        );
      } else {
        inputContainerStyle = ClassNames(styles.input_container, className);
      }
    }

    let innerContainer = ClassNames(
      styles.inner_container,
      innerContainerClassName
    );
    if (IsFocused) {
      if (HasError) {
        innerContainer = ClassNames(
          styles.inner_container,
          styles.failed_regex_focused_label_color,
          innerContainerClassName
        );
      } else {
        innerContainer = ClassNames(
          styles.inner_container,
          styles.focused_input_container,
          innerContainerClassName
        );
      }
    }

    let labelStyle = ClassNames(styles.label, labelClassName);
    if (IsFocused) {
      if (HasError) {
        labelStyle = ClassNames(
          styles.label,
          styles.focused_label,
          styles.failed_regex_focused_label_color,
          labelClassName
        );
      } else {
        labelStyle = ClassNames(
          styles.label,
          styles.focused_label,
          labelClassName
        );
      }
    } else if ((tags && tags.length > 0) || attachments) {
      labelStyle = ClassNames(
        styles.label,
        styles.label_margin,
        labelClassName
      );
    }

    const tagColors = [
      'red',
      'gold',
      'lime',
      'cyan',
      'blue',
      'green',
      'purple',
      'orange',
      'magenta',
      'volcano',
      'geekblue',
    ];

    return (
      <div className={inputContainerStyle}>
        <div className={innerContainer}>
          <div className={styles.content}>
            {label && <p className={labelStyle}>{label}</p>}
            {minLines ? (
              <textarea
                rows={minLines}
                value={Content}
                onBlur={onBlur}
                onFocus={onFocus}
                onChange={onChange}
                placeholder={IsFocused ? hint : ''}
                className={ClassNames(
                  styles.input,
                  inputClassName,
                  labelClassName
                )}
                ref={ref as React.RefObject<HTMLTextAreaElement>}
              />
            ) : (
              <input
                dir="auto"
                type={Type}
                value={Content}
                onBlur={onBlur}
                onFocus={onFocus}
                onChange={onChange}
                placeholder={IsFocused ? hint : ''}
                className={ClassNames(
                  styles.input,
                  inputClassName,
                  labelClassName
                )}
                ref={ref as React.RefObject<HTMLInputElement>}
                onKeyDown={(e) => e.key === 'Enter' && onEnter && onEnter()}
              />
            )}
            {type === 'password' && (
              <div
                data-testid="password-visibility"
                onClick={() =>
                  setType(Type === 'password' ? 'text' : 'password')
                }>
                {Type === 'password' ? (
                  <Assets.SVGs.VisiblePassSvg className={styles.eye_svg} />
                ) : (
                  <Assets.SVGs.InvisiblePassSvg className={styles.eye_svg} />
                )}
              </div>
            )}
            {icon && (
              <div
                onClick={onIconPressed}
                className={styles.eye_svg}
                style={onIconPressed ? {cursor: 'pointer'} : {}}>
                {icon}
              </div>
            )}
          </div>
          {tags && tags.length > 0 && (
            <div
              className={ClassNames(styles.tags_container, styles.scrollview)}>
              {tags.map((tag, i) => (
                <Tag
                  key={tag}
                  closable
                  color={tagColors[i % tagColors.length]}
                  onClose={() => onTagClose && onTagClose(i)}>
                  #{tag}
                </Tag>
              ))}
            </div>
          )}
          {onSelectFile && attachments && (
            <div className={ClassNames(styles.attachments, styles.scrollview)}>
              <label htmlFor="attachment_input">
                <div className={ClassNames(styles.attachment, styles.new)}>
                  <img
                    alt=""
                    src={Assets.Images.Attach}
                    className={styles.attachment_icon}
                  />
                  <p className={styles.name}>چسباندن فایل</p>
                </div>
              </label>
              <input
                type="file"
                id="attachment_input"
                onChange={(e) =>
                  e.target.files &&
                  onSelectFile &&
                  onSelectFile(e.target.files[0])
                }
                className={styles.attachment_input}
              />
              {attachments &&
                attachments.map((attachment, i) => (
                  <div className={styles.attachment} key={i}>
                    <img
                      alt=""
                      src={Assets.Images.Cancel}
                      className={styles.attachment_icon}
                      onClick={() =>
                        onRemoveAttachment && onRemoveAttachment(i)
                      }
                    />
                    <p className={styles.name}>{attachment.name}</p>
                  </div>
                ))}
            </div>
          )}
        </div>
      </div>
    );
  }
);
export default SEInput;
