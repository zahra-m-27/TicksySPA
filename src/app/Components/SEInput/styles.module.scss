$input_primary_color: #0073ff;
$failed_regex_primary_color: #e62e00;

::placeholder {
  color: #B8B8B8;
}

.inner_container {
  direction: rtl;
  border-radius: 8px;
  position: relative;
  box-shadow: 0 0 0 0;
  display: inline-block;
  border: 1px solid #B8B8B8;
}

.focused_input_container {
  border-radius: 8px;
  border-color: $input_primary_color;
  box-shadow: 0 0 8px 0 rgba($input_primary_color, .65);
}

.inner_container {
  display: flex;
  align-items: center;
}

.label {
  top: 50%;
  color: #B8B8B8;
  margin: 0 10px 0;
  transition: 200ms;
  position: absolute;
  pointer-events: none;
  transform: translateY(-50%);
}

.focused_label {
  top: -8px;
  padding: 0 5px;
  margin-right: 8px;
  visibility: visible;
  border-radius: 10px;
  background-color: #fbfbfb;
  color: $input_primary_color;
}

.input {
  height: 48px;
  outline: none;
  border: hidden;
  color: #606060;
  padding: 0 10px;
  font-size: 24px;
  font-weight: 400;
  border-radius: 8px;
  font-family: inherit;
}

.eye_svg {
  display: flex;
  margin-left: 10px;
  align-items: center;
  justify-content: center;
}

.failed_regex_focused_input_container {
  border-color: $failed_regex_primary_color;
  box-shadow: 0 0 8px 0 rgba($failed_regex_primary_color, .65);
}

.failed_regex_focused_label_color {
  color: $failed_regex_primary_color;
}