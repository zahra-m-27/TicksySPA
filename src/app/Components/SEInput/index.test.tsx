import SEInput from '.';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import {render} from '@testing-library/react';
import user from '@testing-library/user-event';

const onChange = jest.fn();

const sample = (
  <SEInput
    name="email"
    label="Test Label"
    content="Test Content"
    onChangeText={onChange}
  />
);
const sample2 = (
  <SEInput label="Test Label" content="Test Content" type="password" />
);
const sample3 = (
  <SEInput
    name="email"
    minLines={5}
    label="Test Label"
    content="Test Content"
    onChangeText={onChange}
  />
);

test('render in ReactDOM', () => {
  const div = document.createElement('div');
  ReactDOM.render(sample, div);
});

test('renders label', () => {
  const {getByText} = render(sample);

  getByText(/Test Label/i);
});

test('renders input value', () => {
  const {getByDisplayValue} = render(sample);

  getByDisplayValue(/Test Content/i);
});

test('renders text area value', () => {
  const {getByDisplayValue} = render(sample3);

  getByDisplayValue(/Test Content/i);
});

test('pass input parameter', () => {
  const {getByDisplayValue} = render(sample);

  const inputElement = getByDisplayValue(/Test Content/i);
  expect(
    inputElement.attributes.getNamedItem('name')?.value === 'email'
  ).toBeTruthy();
});

test('password protection', () => {
  const {getByTestId, getByDisplayValue} = render(sample2);

  const passwordVisibilityButton = getByTestId(/password-visibility/i);

  const input = getByDisplayValue(/Test Content/i);
  expect(input.attributes.getNamedItem('type')?.value).toEqual('password');
  passwordVisibilityButton.click();
  expect(input.attributes.getNamedItem('type')?.value).toEqual('text');
});

it('input onChange', () => {
  const {getByDisplayValue} = render(sample);
  const inputElement = getByDisplayValue(/Test Content/i);
  user.clear(inputElement);
  expect(onChange).toBeCalledWith('', expect.anything());
  user.type(inputElement, 'Test Typing');
  expect(onChange).toBeCalledWith('Test Typing', expect.anything());
});

it('text area onChange', () => {
  const {getByDisplayValue} = render(sample3);
  const inputElement = getByDisplayValue(/Test Content/i);
  user.clear(inputElement);
  expect(onChange).toBeCalledWith('', expect.anything());
  user.type(inputElement, 'Test Typing');
  expect(onChange).toBeCalledWith('Test Typing', expect.anything());
});

it('snapshot Input', () => {
  const snapshot = renderer.create(sample).toJSON();
  expect(snapshot).toMatchSnapshot();
});

it('snapshot TextArea', () => {
  const snapshot = renderer.create(sample3).toJSON();
  expect(snapshot).toMatchSnapshot();
});
