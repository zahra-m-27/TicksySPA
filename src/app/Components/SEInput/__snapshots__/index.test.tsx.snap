// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot Input 1`] = `
<div
  className="input_container"
>
  <div
    className="inner_container"
  >
    <div
      className="content"
    >
      <p
        className="label"
      >
        Test Label
      </p>
      <input
        className="input"
        dir="auto"
        name="email"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        placeholder=""
        type="text"
        value="Test Content"
      />
    </div>
  </div>
</div>
`;

exports[`snapshot TextArea 1`] = `
<div
  className="input_container"
>
  <div
    className="inner_container"
  >
    <div
      className="content"
    >
      <p
        className="label"
      >
        Test Label
      </p>
      <textarea
        className="input"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder=""
        rows={5}
        value="Test Content"
      />
    </div>
  </div>
</div>
`;
